azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java

```diff
@@ -1774,144 +1784,177 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
-        }
-        if (errorMsgs.length() > 0) {
-          // If putting more than 4000 characters in the cookie, the entire
-          // message
-          // will somehow get discarded.
-          registerError(ret, errorMsgs.length() > 4000 ? errorMsgs.substring(0, 4000)
-              : errorMsgs.toString(), resp, 500);
+  private void registerErrorsAndWarningsFromValidationReport(final HttpServletResponse resp,
+      final Map<String, String> ret, final Map<String, ValidationReport> reports) {
+    final StringBuffer errorMsgs = new StringBuffer();
+    final StringBuffer warnMsgs = new StringBuffer();
+    for (final Entry<String, ValidationReport> reportEntry : reports.entrySet()) {
+      final ValidationReport report = reportEntry.getValue();
+      for (final String msg : report.getInfoMsgs()) {
+        switch (ValidationReport.getInfoMsgLevel(msg)) {
+          case ERROR:
+            errorMsgs.append(ValidationReport.getInfoMsg(msg) + "<br/>");
+            break;
+          case WARN:
+            warnMsgs.append(ValidationReport.getInfoMsg(msg) + "<br/>");
+            break;
+          default:
+            break;
         }
-        if (warnMsgs.length() > 0) {
-          ret.put(
-              "warn",
-              warnMsgs.length() > 4000 ? warnMsgs.substring(0, 4000) : warnMsgs
-                  .toString());
-        }
-      } catch (final Exception e) {
-        logger.info("Installation Failed.", e);
-        String error = e.getMessage();
-        if (error.length() > 512) {
-          error =
-              error.substring(0, 512) + "<br>Too many errors to display.<br>";
-        }
-        registerError(ret, "Installation Failed.<br>" + error, resp, 500);
-      } finally {
-        if (out != null) {
-          out.close();
+      }
+      if (!report.getErrorMsgs().isEmpty()) {
+        errorMsgs.append("Validator " + reportEntry.getKey() + " reports errors:<ul>");
+        for (final String msg : report.getErrorMsgs()) {
+          errorMsgs.append("<li>" + msg + "</li>");
         }
-        if (tempDir.exists()) {
-          FileUtils.deleteDirectory(tempDir);
+        errorMsgs.append("</ul>");
+      }
+      if (!report.getWarningMsgs().isEmpty()) {
+        warnMsgs.append("Validator " + reportEntry.getKey() + " reports warnings:<ul>");
+        for (final String msg : report.getWarningMsgs()) {
+          warnMsgs.append("<li>" + msg + "</li>");
         }
+        warnMsgs.append("</ul>");
       }
-
-      logger.info("Upload: project " + projectName + " version is " + project.getVersion()
-          + ", reference is " + System.identityHashCode(project));
-      ret.put("version", String.valueOf(project.getVersion()));
+    }
+    if (errorMsgs.length() > 0) {
+      // If putting more than 4000 characters in the cookie, the entire message will somehow
+      // get discarded.
+      registerError(ret,
+          errorMsgs.length() > 4000 ? errorMsgs.substring(0, 4000) : errorMsgs.toString(), resp,
```

Comment:
Is there a reason why this is 4000 and not 512? Can this be consistent?

Reply:
I'm only moving the code around to reduce the size of ajaxHandleUpload(). I don't know what these numbers: 512 and 4000 represent.

Reply:
I think this is a good chance to understand why the numbers are different. Can you please find out?

Reply:
512 limits the size of an unexpected error message while processing the request and seems a pretty arbitrary choice.
4000 limits the size of project validation messages, apparently it has something to do with browser cookies, but if that's the case the right place to do the trimming is when it's actually stored in a cookie, not here.
This is the only place where these numbers are used and we wouldn't be gaining much by making this change. Given our current priorities I would suggest to leave it as it is and revisit this when we have a better motivation.